geoutils/filters.py

```diff
@@ -0,0 +1,213 @@
+    """
+    Apply a filter to the array. See description of raster.filter.
+    """
+    if isinstance(method, str):
+        filter_map: dict[str, Callable[..., NDArrayNum]] = {
+            "gaussian": gaussian_filter,
+            "median": median_filter,
+            "mean": mean_filter,
+            "max": max_filter,
+            "min": min_filter,
+            "distance": distance_filter,
+        }
+
+        if method not in filter_map:
+            raise ValueError(f"Unsupported filter method '{method}'. " f"Available methods: {list(filter_map.keys())}")
+
+        filter_func = filter_map[method]
+
+    elif callable(method):
+        filter_func = method
+    else:
+        raise TypeError("`method` must be a string or a callable.")
+
+    # Apply filter
+    filtered_data = filter_func(array, **kwargs)
+    return filtered_data
+
+
+def _nan_safe_filter(array: NDArrayNum, filter_func: Callable[..., NDArrayNum], **kwargs: Any) -> NDArrayNum:
+    """Apply a NaN-safe filter using a weighting trick."""
+    # Check that array dimension is 2 or 3
+    if np.ndim(array) not in [2, 3]:
+        raise ValueError(f"Invalid array shape given: {array.shape}. Expected 2D or 3D array.")
+
+    # In case array does not contain NaNs, use scipy's gaussian filter directly
+    if not np.isnan(array).any():
+        return filter_func(array, **kwargs)
+
+    # If array contain NaNs, need a more sophisticated approach
+    # Inspired by https://stackoverflow.com/a/36307291
+    # Run filter on the array with NaNs set to 0
+    array_filled = np.nan_to_num(array, nan=0.0)
+    weights = (~np.isnan(array)).astype(array_filled.dtype)
+
+    filtered_array_filled = filter_func(array_filled, **kwargs)
+    del array_filled
+
+    weight_sum = filter_func(weights, **kwargs)
+    del weights
+
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", message="invalid value encountered")
+        filtered_array = filtered_array_filled / weight_sum
+
+    return filtered_array
+
+
+def gaussian_filter(array: NDArrayNum, sigma: float) -> NDArrayNum:
+    """
+    Apply a Gaussian filter to a raster that may contain NaNs, using scipy's implementation.
+    gaussian_filter_cv is recommended as it is usually faster, but this depends on the value of sigma.
+
+    N.B: kernel_size is set automatically based on sigma.
+
+    :param array: the input array to be filtered.
+    :param sigma: the sigma of the Gaussian kernel
+
+    :returns: the filtered array (same shape as input)
+    """
+    return _nan_safe_filter(array, scipy.ndimage.gaussian_filter, sigma=sigma)
+
+
+def median_filter(array: NDArrayNum, **kwargs: Any) -> NDArrayNum:
+    """
+    Apply a median filter to a raster that may contain NaNs, using scipy's implementation.
+
+    :param array: the input array to be filtered.
+
+    :returns: the filtered array (same shape as input).
+    """
+    return _nan_safe_filter(array, scipy.ndimage.median_filter, **kwargs)
```

Comment:
For the median we need to keep the original implementation that used `np.inf`, which are ignored by `median`.
See the behaviour:
```python
import numpy as np
array = [1, 2, 3]
np.median(array)
Out[4]: np.float64(2.0)
array2 = [1, 2, 3, np.inf, np.inf, np.inf, np.inf]
np.median(array)
Out[6]: np.float64(2.0)
```
I think replacing by zeros in `_nan_safe_filter` won't work:
```python
array = [1, 2, 3, 0, 0, 0, 0]
np.median(array)
Out[8]: np.float64(0.0)
```
Because no matter what you divide the 0 by above, it will never be equal to the real result 2.

Reply:
Should we simply call `generic_filter` with `nanmean` and `nanmedian` inside?

Reply:
This would be the simplest solution, but from my experience `generic_filter` is really slow. So we might consider other options in the future. At least, we could compare our implementation to that in the tests.

Reply:
Yes that's unfortunately the case, see here (46 times slower to use `generic_filter` for `median`): https://scipy-user.scipy.narkive.com/YyMHNgYg/scipy-ndimage-filters-median-filter-with-nan-values

The other option is to use Numba as default, which should be fast for any custom operation like this.
This would mirror the behaviour of xDEM in terrain attributes, where I'm adding the option of using either a Numba or Scipy engine.

For this PR, we can stick to SciPy only and `generic_filter`.

Reply:
There's also a way to write a short C filter to pass to SciPy, which is usually much faster.
See: http://notmatthancock.github.io/2020/04/19/scipy-generic-filter-low-level-callable.html

Reply:
Actually my bad, for the `nanmean` we had a working filter already using the trick to replace by the zeros, and then dividing by the sum of valid pixels only, but I don't think this how it behaves exactly in the implementation of this PR.
We actually have an old working duplicate here: https://github.com/GlacioHack/xdem/blob/e1e2065ea52a7db3c8e88045d743aa1f6efd08bb/xdem/spatialstats.py#L2563

In short, to move forward in this PR, we need to remove the `_nan_safe_filter()` function (as this trick only works for the mean, it's not generalizable), re-implement the mean only with this trick.
And, for the median, simply use `generic_filter(np.nanmedian)` (which is quite slow, but there are no other choices I think).